http-clients/apache-client/src/test/java/software/amazon/awssdk/http/apache/internal/RepeatableInputStreamRequestEntityTest.java

```diff
@@ -0,0 +1,799 @@
+    }
+
+    @Test
+    @DisplayName("writeTo should reset stream on subsequent attempts if repeatable")
+    void writeTo_SubsequentAttemptWithRepeatableStream_ResetsStream() throws IOException {
+        // Given
+        String content = "test content";
+
+        // Create a custom stream that tracks reset calls
+        AtomicInteger resetCallCount = new AtomicInteger(0);
+        ByteArrayInputStream trackingStream = new ByteArrayInputStream(content.getBytes()) {
+            @Override
+            public synchronized void reset()  {
+                resetCallCount.incrementAndGet();
+                super.reset();
+            }
+        };
+
+        ContentStreamProvider provider = () -> trackingStream;
+
+        SdkHttpRequest httpRequest = httpRequestBuilder.build();
+        HttpExecuteRequest request = HttpExecuteRequest.builder()
+                                                       .request(httpRequest)
+                                                       .contentStreamProvider(provider)
+                                                       .build();
+
+        entity = new RepeatableInputStreamRequestEntity(request);
+
+        // First write
+        ByteArrayOutputStream firstOutput = new ByteArrayOutputStream();
+        entity.writeTo(firstOutput);
+
+        //Second write
+        ByteArrayOutputStream secondOutput = new ByteArrayOutputStream();
+        entity.writeTo(secondOutput);
+
+        // Then
+        assertEquals(content, firstOutput.toString());
+        assertEquals(content, secondOutput.toString());
+        assertEquals(1, resetCallCount.get(), "Reset should be called exactly once for second attempt");
+    }
+
+    @Test
+    @DisplayName("writeTo should preserve original exception on first failure")
+    void writeTo_FirstAttemptThrowsException_PreservesOriginalException() throws IOException {
+        // Given
+        IOException originalException = new IOException("Original error");
+        InputStream faultyStream = mock(InputStream.class);
+        when(faultyStream.read(any(byte[].class))).thenThrow(originalException);
+        when(faultyStream.markSupported()).thenReturn(true);
+
+        ContentStreamProvider provider = () -> faultyStream;
+        SdkHttpRequest httpRequest = httpRequestBuilder.build();
+        HttpExecuteRequest request = HttpExecuteRequest.builder()
+                                                       .request(httpRequest)
+                                                       .contentStreamProvider(provider)
+                                                       .build();
+
+        entity = new RepeatableInputStreamRequestEntity(request);
+
+        IOException thrown = assertThrows(IOException.class,
+                                          () -> entity.writeTo(new ByteArrayOutputStream()));
+        assertSame(originalException, thrown);
+    }
+
+    @Test
+    @DisplayName("writeTo should throw original exception on subsequent failures")
```

Comment:
why do we want to preserve the original exception?

Reply:
For better user experience , since this is mark and reset in cases of retry the stream will be restarted and if it fails the user need to know the initial error that caused the retry.